File: 76. Minimum Window Substring.py
class Solution(object):

    def minWindow(self, s, t):

        """

        :type s: str

        :type t: str

        :rtype: str

        """

        start, end, left = 0, 0, 0

        min_len = sys.maxsize

        hash = collections.defaultdict(int)

        for i in t:

            hash[i] += 1

        num = len(t)

        for i, x in enumerate(s):

            if(hash[x]>0):

                num -= 1

            hash[x] -= 1

            while(num==0):

                if (i-left +1) < min_len:

                    start = left

                    min_len = i - left + 1

                if hash[s[left]]==0:

                    num += 1

                hash[s[left]]+=1

                left += 1

        if min_len== sys.maxsize:

            return ""

        else:

            return s[start:start+min_len]

        

File: 76. Minimum Window Substring.cpp
class Solution2 {
public:
    string minWindow(string s, string t) {
        int sLen = s.size(), tLen = t.size();
        if(sLen < tLen || tLen == 0) return "";
        vector<int> hash(256, 0);
        for(auto& ch : t)   ++hash[static_cast<int>(ch)];
        int minLen = INT_MAX, start = 0, minStart = 0, i = 0, found = 0;
        while(i <= sLen && start < sLen){
            if(found < tLen){
                if(i == sLen)   break;
                --hash[s[i]];
                if(hash[s[i++]] >= 0) ++found;
            }
            else{
                if(i - start < minLen){
                    minLen = i - start;
                    minStart = start;
                }
                hash[s[start]]++;
                if(hash[s[start++]] > 0)    --found;
            }
        }
        return minLen <= sLen ? s.substr(minStart, minLen) : "";
    }
};